WIP initial commit of changes, next step is unit tests

brbrown25 opened the pull request:
https://issues.redhat.com/projects/DBZ/issues/DBZ-1813

@gunnarmorling I havn't implemented the tests yet, but wanted to run this by you first before embarking on that. If you like this implementation I can work on adding in some unit tests that show the different modes.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -592,6 +592,84 @@ public static SchemaRefreshMode parse(String value) {
+        /**
+         * Enable publication on a specific set of tables.
+         */
+        WHITELIST("whitelist");
```
Actually it's whitelist + blacklist together. Perhaps "filter" or "filtered"?

brbrown25 replied:
yeah filtered makes the most sense I think given the desired behaviour

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:
```diff
@@ -111,6 +111,8 @@ protected ReplicationConnection createReplicationConnection(boolean exportSnapsh
         return ReplicationConnection.builder(config.jdbcConfig())
                 .withSlot(config.slotName())
                 .withPublication(config.publicationName())
+                .withTableWhitelist(config.tableWhitelist())
```
This needs to consider also the table blacklist as well as database blacklist/whitelist. I suppose this was always an implicit assumption on my end during the conversations, sorry for not being more clear on that. You can take a look at `RelationalSnapshotChangeEventSource.determineCapturedTables(RelationalSnapshotContext)` to see how that's handled during snapshotting using the `TableFilter` contract.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/ReplicationConnection.java:
```diff
@@ -133,6 +133,24 @@ static String format(long lsn) {
+         * @return this instance
+         * @see #config.tableWhitelist()
+         */
+        Builder withTableWhitelist(final String tableWhitelist);
```
As per above, this probably should receive `TableFilter` instead.

brbrown25 commented:
Sweet I’ll work that out over the weekend and update the Pr!

brbrown25 commented:
@gunnarmorling working on this changes and should have something in the next day or two!

brbrown25 commented:
@gunnarmorling I'm sure there is a better way to doing the tablefilter portion, so any guidance is greatly appreciated!

gunnarmorling commented:
@brbrown25, there's quite a few formatting changes/issues. Can you please run `mvn process-sources` to apply the auto-formatter? Thanks!

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -117,38 +136,97 @@ private PostgresReplicationConnection(Configuration config,
-                                    "A logical publication named '{}' for plugin '{}' and database '{}' is already active on the server " +
-                                            "and will be used by the plugin",
-                                    publicationName, plugin, database());
+            if (PostgresConnectorConfig.AutoCreateMode.DISABLED.equals(publicationAutocreateMode)) {
```
Shouldn't this be handled further below, in the swich block? It seems we'd trigger this exception in case the publication already exists?

brbrown25 replied:
I went for a failing early approach, but I can see the argument for moving it into the switch, now that I look at it, that seems like a better place for it!

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -117,38 +136,97 @@ private PostgresReplicationConnection(Configuration config,
+        }
+    }
+
+    private Stream<TableId> toTableIds(Set<TableId> tableIds, Pattern pattern) {
```
The name is a bit misleading, perhaps simply `filter()`?

brbrown25 replied:
yeah this was copied from the `RelationalSnapshotChangeEventSource.determineCapturedTables(RelationalSnapshotContext)`

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -117,38 +136,97 @@ private PostgresReplicationConnection(Configuration config,
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+    }
+
+    private Set<TableId> determineCapturedTables(RelationalSnapshotChangeEventSource.RelationalSnapshotContext ctx) throws Exception {
```
Why is the context parameter needed here?

brbrown25 replied:
that's the part where I wasn;t really sure the best way of using the table filter contract. since based on `RelationalSnapshotChangeEventSource.determineCapturedTables(RelationalSnapshotContext)` it appears that is needs to know all the tables and then iterate through and build up a list of TableIds that match

brbrown25 replied:
actually I may have just figure out how that works...should have  some updates later today! thanks so much for helping me through on this!

gunnarmorling commented on debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:
```diff
@@ -218,8 +219,7 @@ public void shouldSupportSSLParameters() throws Exception {
             assertConnectorNotRunning();
-        }
-        else {
+        } else {
```
Would be better to avoid those formatting changes. The formatter will flag/revert them anyways, and if it doesn't, it adds some unneccessary noise to the PR.

brbrown25 replied:
> Hey @brbrown25, a few comments inline. Overall I think it's going into the right direction. Having a test which asserts some events from a filtered publication would be nice to have, too. Plus, the new option should be added to the connector docs. As a heads-up, we'll do the next 1.2 release mid next week. Looks like we could have this wrapped up by then?
> 
> Thanks a lot!

Yeah I think we can have it wrapped up by then!

brbrown25 commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -117,38 +136,97 @@ private PostgresReplicationConnection(Configuration config,
+                .sorted();
+    }
+
+    private Set<TableId> sort(Set<TableId> capturedTables) throws Exception {
```
So this was taken from the the `RelationalSnapshotChangeEventSource.determineCapturedTables(RelationalSnapshotContext)` so I copied that over for consistency

brbrown25 commented:
- [ ] update connector docs
- [ ] Add events filtered from publication

brbrown25 commented:
@gunnarmorling I’m gonna start the docs Pr this evening and will work out the additional test to show filtered events as well. I was able to remove the context once I figured out things (thanks for the help!). Also the build is breaking but when I’m getting the same failure in the master branch.

gunnarmorling commented:
@brbrown25, thanks for all the updates; could you ping me once it's done from your side?

brbrown25 commented:
> @brbrown25, thanks for all the updates; could you ping me once it's done from your side?

It’s good for you to review just let me know any changes and I can make them! Also was wondering is the documentation supposed to be a Pr in the github.io repo as well? Thanks!

gunnarmorling commented:
> @gunnarmorling It’s good for you to review just let me know any changes and I can make them! 

Thanks, going to re-review now.

> Also was wondering is the documentation supposed to be a Pr in the github.io repo as well?

Nope, just here in the main repo is all that's needed (i.e. what you already did). The website build pulls in the docs from there for the different branches.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -125,12 +144,29 @@ protected void initPublication() {
+                                        stmt.execute(createPublicationStmt);
+                                    }
+                                    catch (Exception e) {
+                                        LOGGER.error("Unable to create publication for ");
```
There seems to be missing something after "for". More importantly, though, did you consciously decide to not propagate that exception? I think it'd be better to just let it bubble up (wrapped in a `ConnectException` with some contextual information), thus stopping connector start-up.

brbrown25 replied:
Actually it looks like I messed something up, I do want to through a connect exception and provide a better error message!

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -151,6 +187,42 @@ protected void initPublication() {
+            }
+        }
+
+        if (value != null) {
```
Hum, I think I've asked about it before, but why is this explicit check against TABLE_WHITELIST done here? It shouldn't be needed, given you already apply the data collection filter above. Is it some sort of left-over perhaps?

brbrown25 replied:
yeah this does seem like a left over!

gunnarmorling commented on debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:
```diff
@@ -1628,6 +1630,126 @@ public void shouldOutputRecordsInCloudEventsFormat() throws Exception {
+
+        Configuration.Builder configBuilder = TestHelper.defaultConfig()
+                .with(PostgresConnectorConfig.PUBLICATION_NAME, "cdc")
+                .with(PostgresConnectorConfig.TABLE_WHITELIST, "public.numeric_table")
```
Could you change the test so that multiple tables are matched? Seems like a critical thing to get right, set up the publication for more than one table.

brbrown25 replied:
will do! seems like a good thing to have!

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/postgresql.adoc:
```diff
@@ -1892,6 +1892,18 @@ See the {link-prefix}:{link-postgresql-connector}#postgresql-data-types[list of
 
+|[[postgresql-publication-autocreate-mode]]<<postgresql-publication-autocreate-mode, `publication.autocreate.mode`>>
+|_all_tables_
+|Determine how creation of a publication should work, the default is all_tables.
```
Could you clarify that this option only applies when pgoutput is used? Also, perhaps a link to upstream Postgres docs on the concept of publications?

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/postgresql.adoc:
```diff
@@ -1892,6 +1892,18 @@ See the {link-prefix}:{link-postgresql-connector}#postgresql-data-types[list of
+_disabled_ - The connector will not attempt to create a publication at all. The expectation is
+that the user has created the publication up-front. If the publication isn't found to exist upon
+startup, the connector will throw an exception and stop.
+_all_tables_ - This acts like the current behavior where if the publication does not exist the
```
Let's avoid the notion of "current" behavior, as it won't be clear to the future reader:

> If no publication exists, the connector will create a new publication for all tables. Note this requires xyz grants. If the publication already exists, it will be used.

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/postgresql.adoc:
```diff
@@ -1892,6 +1892,18 @@ See the {link-prefix}:{link-postgresql-connector}#postgresql-data-types[list of
+_all_tables_ - This acts like the current behavior where if the publication does not exist the
+connector will create the publication for all tables. If the publication exists, we will use it.
+i.e `CREATE PUBLICATION <publication_name> FOR ALL TABLES;`
+_filtered_ - This would create the publication for matching tables only
```
If no publication exists, the connector will create a new publication for all those tables matching the current filter configuration (see table/database whitelist/blacklist properties). If the publication already exists, it will be used.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:
```diff
@@ -592,6 +592,83 @@ public static SchemaRefreshMode parse(String value) {
+                            "DISABLED - The connector will not attempt to create a publication at all. The expectation is " +
+                            "that the user has created the publication up-front. If the publication isn't found to exist upon " +
+                            "startup, the connector will throw an exception and stop." +
+                            "ALL_TABLES - This acts like the current behavior where if the publication does not exist the " +
```
Same comment on "current" as further below.

brbrown25 commented:
@gunnarmorling I've gone ahead and collapsed this down into a single clean commit. I've addressed all of the feedback (thanks so much!) if there are any other tweaks, I can make them. Really excited for this feature!

brbrown25 commented:
darn looks like i broke a test, well fix that shortly!

gunnarmorling commented:
> darn looks like i broke a test, well fix that shortly!

There are a few flaky tests. Do the two failing ones pass for you locally?

brbrown25 commented:
> There are a few flaky tests. Do the two failing ones pass for you locally?
I'm running it now, to verify, when I did it in intellij `shouldProduceEventsWhenAlwaysTakingSnapshots` would fail when running the whole suite, but running just that test was successful, so i'm running on commandline now to see what happens.

brbrown25 commented:
@gunnarmorling confirmed that locally I now get
```
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   RecordsSnapshotProducerIT.shouldCloseTransactionsAfterSnapshot:286 Values should be different. Actual: idle in transaction
[ERROR] Errors:
[ERROR]   PostgresConnectorIT.shouldProduceEventsWhenAlwaysTakingSnapshots:640->assertRecordsFromSnapshot:1809 NullPointer
```

gunnarmorling commented:
Interesting. So I'd suggest you try and a) see whether those tests pass on current master (might have merged something broken accidentally) and if they do, b) to rebase the PR on current master (a good idea anyways) and see how things then look.

brbrown25 commented:
> Interesting. So I'd suggest you try and a) see whether those tests pass on current master (might have merged something broken accidentally) and if they do, b) to rebase the PR on current master (a good idea anyways) and see how things then look.

ok indeed my most recent commit broke something, it's in sync with master and master passes for me. will get this worked out this evening and ping!

brbrown25 commented:
@gunnarmorling master passes for me, i think i may have been out of sync on my branch, checking one more time

gunnarmorling commented:
The remaining test failure is unrelated. Pushed one commit which hopefully gets rid of it.

brbrown25 commented:
> The remaining test failure is unrelated. Pushed one commit which hopefully gets rid of it.

Thanks so much! It's always a joy contributing to Debezium, so thanks for all of the help on this.

brbrown25 commented:
looks like the build is green, yay!

gunnarmorling commented:
yeah, indeed. Waiting for @jpechane to do one last review pass and merge. Thanks for your efforts and contribution, @brbrown25!

gunnarmorling commented:
Woohoo!

>
